onedal/datatypes/utils/sua_iface_helpers.cpp

```diff
@@ -0,0 +1,225 @@
+// which is a Python integer encoding USM pointer value.
+std::uintptr_t get_sua_ptr(const py::dict& sua) {
+    const py::tuple data = get_sua_data(sua);
+    return data[0ul].cast<std::uintptr_t>();
+}
+
+// Get `__sycl_usm_array_interface__['data'][1]`, which is the data entry a read-only
+// flag (True means the data area is read-only).
+bool is_sua_readonly(const py::dict& sua) {
+    const py::tuple data = get_sua_data(sua);
+    return data[1ul].cast<bool>();
+}
+
+// Get `__sycl_usm_array_interface__['shape']`.
+// shape : a tuple of integers describing dimensions of an N-dimensional array.
+py::tuple get_sua_shape(const py::dict& sua) {
+    py::tuple shape = sua["shape"].cast<py::tuple>();
+    if (shape.size() == py::ssize_t{ 0ul }) {
+        throw std::runtime_error("Wrong number of dimensions");
+    }
+    return shape;
+}
+
+void report_problem_for_sua_iface(const char* clarification) {
+    constexpr const char* const base_message = "Unable to convert from SUA interface";
+
+    std::string message{ base_message };
+    message += std::string{ clarification };
+    throw std::invalid_argument{ message };
+}
+
+// Get and check `__sycl_usm_array_interface__` number of dimensions.
+std::int64_t get_and_check_sua_iface_ndim(const py::dict& sua_dict) {
+    constexpr const char* const err_message = ": only 1D & 2D tensors are allowed";
+    py::tuple shape = get_sua_shape(sua_dict);
+    const py::ssize_t raw_ndim = shape.size();
+    const auto ndim = detail::integral_cast<std::int64_t>(raw_ndim);
+    if ((ndim != 1l) && (ndim != 2l))
+        report_problem_for_sua_iface(err_message);
+    return ndim;
+}
+
+// Get the pair of row and column counts.
+std::pair<std::int64_t, std::int64_t> get_sua_iface_shape_by_values(const py::dict sua_dict,
+                                                                    const std::int64_t ndim) {
+    std::int64_t row_count, col_count;
+    auto shape = sua_dict["shape"].cast<py::tuple>();
+    if (ndim == 1l) {
+        row_count = shape[0l].cast<std::int64_t>();
+        col_count = 1l;
+    }
+    else {
+        row_count = shape[0l].cast<std::int64_t>();
+        col_count = shape[1l].cast<std::int64_t>();
+    }
+    return std::make_pair(row_count, col_count);
+}
+
+// Get oneDAL Homogen DataLayout enumeration from input object shape and strides.
+dal::data_layout get_sua_iface_layout(const py::dict& sua_dict,
+                                      const std::int64_t& r_count,
+                                      const std::int64_t& c_count) {
+    const auto raw_strides = sua_dict["strides"];
+    if (raw_strides.is_none()) {
+        // None to indicate a C-style contiguous array.
+        return dal::data_layout::row_major;
+    }
+    auto strides_tuple = raw_strides.cast<py::tuple>();
+
+    auto strides_len = py::len(strides_tuple);
+
+    if (strides_len == 1l) {
+        return dal::data_layout::row_major;
+    }
+    else if (strides_len == 2l) {
```

Comment:
Such inputs are not expected.
All input array, numpy or non-numpy should be validated before passing them to these primitives.

Reply:
Again there is no heterogenous table support yet here. Please take a look once again the description and the code here.

Reply:
@icfaust I assume that what was meant here was non-contiguous array, not heterogeneous tables.
This terminology just caused some confusion in my interpretation of the comments.

On main and in this PR branch, it is expected that input data for `to_table` should be contiguous, the primitive should be called after such input data validators as` validate_data`, `check_array` and etc.

Based on this, it might be worth updating and making the error call consistent for numpy ndarrays and usm_ndarrays, and adding tests accordingly.